chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+        learning_rate - search for max_lr, init_lr, final_lr, and warmup_epochs. The search for init_lr and final_lr values
+            are defined as fractions of the max_lr value. The search for warmup_epochs is as a fraction of the total epochs used.
+        all - include search for all 13 inidividual keyword options
+
+    Individual supported parameters:
+        {get_available_spaces(0).keys()}
+    """,
+    )
+
+    hyperopt_args.add_argument(
+        "--n-cpu-per-worker",
+        type=int,
+        default=0,
+        help="Number of CPUs to allocate for each worker",
+    )
+
+    hyperopt_args.add_argument(
+        "--n-gpu-per-worker",
+        type=int,
+        default=0,
+        help="Number of GPUs to allocate for each worker",
+    )
+
+    hyperopt_args.add_argument(
+        "--num-checkpoints-to-keep",
+        type=int,
+        default=1,
+        help="Number of checkpoints to keep",
+    )
+
+    return parser
+
+def process_hyperopt_args(args: Namespace) -> Namespace:
+    if args.startup_random_iters is None:
+        args.startup_random_iters = args.num_samples // 2
+
+    if args.hyperopt_save_dir is None:
+        args.hyperopt_save_dir = Path(f"chemprop_hyperopt/{args.data_path.stem}")
+
+    args.hyperopt_save_dir.mkdir(exist_ok=True, parents=True)
+
+    search_parameters = set()
+
+    for keyword in args.search_parameter_keywords:
+        if keyword not in SEARCH_PARAM_KEYWORDS_MAP and keyword not in AVAILABLE_SPACES:
+            raise ValueError(f"Invalid search parameter keyword: {keyword}")
+        
+        search_parameters.update(SEARCH_PARAM_KEYWORDS_MAP[keyword] if keyword in SEARCH_PARAM_KEYWORDS_MAP else [keyword])
+
+    args.search_parameter_keywords = list(search_parameters)
+
+    return args
+
+def build_search_space(search_parameters: list[str], train_epochs: int) -> dict:
+    AVAILABLE_SPACES = get_available_spaces(train_epochs)
+
+    return {param: AVAILABLE_SPACES[param] for param in search_parameters}
+
+def update_args_with_config(args: Namespace, config: dict) -> Namespace:
+    for key, value in config.items():
+
+        match key:
+
+            case "linked_hidden_size":
+                setattr(args, "hidden_size", value)
+                setattr(args, "ffn_hidden_size", value)
+
+            case "final_lr_ratio":
+                setattr(args, "final_lr", value * args.max_lr)
+
+            case "init_lr_ratio":
+                setattr(args, "init_lr", value * args.max_lr)
+
+            case _:
+                setattr(args, key, value)
+
+    return args
+
+def train_model(config, args, train_loader, val_loader, logger):
+
+    update_args_with_config(args, config)
+
+    model = build_model(args, train_loader.dataset)
+    logger.info(model)
+
+    monitor_mode = "min" if model.metrics[0].minimize else "max"
+    logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+
+    trainer = pl.Trainer(
```

Comment:
We should consider merging https://github.com/chemprop/chemprop/pull/655 before this PR to simplify the calls to `Trainer`